pimd/pim6_mld.c

```diff
@@ -0,0 +1,2705 @@
+static const pim_addr gm_dummy_untracked = { .s_addr = 0xffffffff, };
+#endif
+/* clang-format on */
+
+#define IPV6_MULTICAST_SCOPE_LINK 2
+
+static inline uint8_t in6_multicast_scope(const pim_addr *addr)
+{
+	return addr->s6_addr[1] & 0xf;
+}
+
+static inline bool in6_multicast_nofwd(const pim_addr *addr)
+{
+	return in6_multicast_scope(addr) <= IPV6_MULTICAST_SCOPE_LINK;
+}
+
+/*
+ * (S,G) -> subscriber,(S,G)
+ */
+
+static int gm_packet_sg_cmp(const struct gm_packet_sg *a,
+			    const struct gm_packet_sg *b)
+{
+	const struct gm_packet_state *s_a, *s_b;
+
+	s_a = gm_packet_sg2state(a);
+	s_b = gm_packet_sg2state(b);
+	return IPV6_ADDR_CMP(&s_a->subscriber->addr, &s_b->subscriber->addr);
+}
+
+DECLARE_RBTREE_UNIQ(gm_packet_sg_subs, struct gm_packet_sg, subs_itm,
+		    gm_packet_sg_cmp);
+
+static struct gm_packet_sg *gm_packet_sg_find(struct gm_sg *sg,
+					      enum gm_sub_sense sense,
+					      struct gm_subscriber *sub)
+{
+	struct {
+		struct gm_packet_state hdr;
+		struct gm_packet_sg item;
+	} ref = {
+		/* clang-format off */
+		.hdr = {
+			.subscriber = sub,
+		},
+		.item = {
+			.offset = 0,
+		},
+		/* clang-format on */
+	};
+
+	return gm_packet_sg_subs_find(&sg->subs[sense], &ref.item);
+}
+
+/*
+ * interface -> (*,G),pending
+ */
+
+static int gm_grp_pending_cmp(const struct gm_grp_pending *a,
+			      const struct gm_grp_pending *b)
+{
+	return IPV6_ADDR_CMP(&a->grp, &b->grp);
+}
+
+DECLARE_RBTREE_UNIQ(gm_grp_pends, struct gm_grp_pending, itm,
+		    gm_grp_pending_cmp);
+
+/*
+ * interface -> ([S1,S2,...],G),pending
+ */
+
+static int gm_gsq_pending_cmp(const struct gm_gsq_pending *a,
+			      const struct gm_gsq_pending *b)
+{
+	if (a->s_bit != b->s_bit)
+		return numcmp(a->s_bit, b->s_bit);
+
+	return IPV6_ADDR_CMP(&a->grp, &b->grp);
+}
+
+static uint32_t gm_gsq_pending_hash(const struct gm_gsq_pending *a)
+{
+	uint32_t seed = a->s_bit ? 0x68f0eb5e : 0x156b7f19;
+
+	return jhash(&a->grp, sizeof(a->grp), seed);
+}
+
+DECLARE_HASH(gm_gsq_pends, struct gm_gsq_pending, itm, gm_gsq_pending_cmp,
+	     gm_gsq_pending_hash);
+
+/*
+ * interface -> (S,G)
+ */
+
+static int gm_sg_cmp(const struct gm_sg *a, const struct gm_sg *b)
+{
+	return pim_sgaddr_cmp(a->sgaddr, b->sgaddr);
+}
+
+DECLARE_RBTREE_UNIQ(gm_sgs, struct gm_sg, itm, gm_sg_cmp);
+
+static struct gm_sg *gm_sg_find(struct gm_if *gm_ifp, pim_addr grp,
+				pim_addr src)
+{
+	struct gm_sg ref;
+
+	ref.sgaddr.grp = grp;
+	ref.sgaddr.src = src;
+	return gm_sgs_find(gm_ifp->sgs, &ref);
+}
+
+static struct gm_sg *gm_sg_make(struct gm_if *gm_ifp, pim_addr grp,
+				pim_addr src)
+{
+	struct gm_sg *ret, *prev;
+
+	ret = XCALLOC(MTYPE_GM_SG, sizeof(*ret));
+	ret->sgaddr.grp = grp;
+	ret->sgaddr.src = src;
+	ret->iface = gm_ifp;
+	gm_packet_sg_subs_init(ret->subs_positive);
+	gm_packet_sg_subs_init(ret->subs_negative);
+	prev = gm_sgs_add(gm_ifp->sgs, ret);
+
+	if (prev) {
+		gm_packet_sg_subs_init(ret->subs_positive);
+		gm_packet_sg_subs_init(ret->subs_negative);
```

Comment:
The above 2 lines are redundant, can be removed.

Reply:
Ah, no, it's a typo, it should be `_fini` … if `gm_sgs_add` returns non-NULL, a preexisting entry was found. Will clean/fix.